### ISSUE TITLE
LAPS silently returns a frozen, un-equilibrated ensemble after a single-chain NaN (~13% of runs)

### ISSUE THREAD

**Issue by junpenglao** [blackjax-devs/blackjax#1035]
## Summary

`laps()` can return a **frozen, un-equilibrated ensemble with no error, no warning, and no NaN in the returned positions**. The draws look plausible and mean nothing. Measured rate: **~13% of runs** at the configuration used by `tests/mcmc/test_sampling.py::LinearRegressionTest::test_laps`.

This is a pre-existing defect, not a recent regression. It surfaced because the JAX Nightly CI began failing `test_laps` deterministically from 2026-09-04 — but the nightly did not break anything. A **+1 ULP perturbation of `y_data`** at the test's own fixed seed, on **pinned** JAX 0.10.0, reproduces the CI value to within float32 round-off:

| | `np.mean(np.exp(final_state.position[:, 0]))` |
|---|---|
| CI nightly, six runs, bit-identical | `1.963122` |
| pinned JAX 0.10.0, `y_data` +1 ULP | `1.9631262` |
| pinned JAX 0.10.0, unperturbed | `1.0320567` |
| ground truth (independent 4-chain NUTS) | `1.033311` |

The nightly re-rolled a die this code has been rolling all along.

## Mechanism

Traced on the `y_ulp=+1` run; ensemble means over 100 chains per burn-in iteration, `_nanfrac` being the kernel's own NaN verdict:

```
iter    _nanfrac  _ld_finite  _kc_isnan  _ec_isnan        _E      EEVPD  step_size
  21           0           1          0          0   1.33547    5.59801   0.134128
  22           0           1       0.01       0.01       nan        nan        nan
  23           1           1          0          0         0          0        nan
  24           1           1          0          0         0          0        nan
```

1. **One chain in 100 produces a NaN `kinetic_change`** while its position, momentum and logdensity are all finite. `handle_nans` (`blackjax/mcmc/mclmc.py:229-236`) checks only position, momentum and logdensity, so `info.nonans` is `True` and the kernel reports a clean step. (`energy_change = kinetic_change - logdensity + state.logdensity`, `mclmc.py:124`; the NaN route is `jnp.log(1 + momentum_proj + (1 - momentum_proj) * zeta**2)` at `integrators.py:472-473`, where `momentum_proj` is a dot product of unit vectors that can round marginally below `-1` in float32.)
2. **That one chain poisons all 100.** ECA averages summary statistics with `lax.psum` (`blackjax/eca.py:52-59`), so `Etheta["E"]`/`Etheta["Esq"]` go NaN for the whole ensemble → `EEVPD` NaN (`laps_burn_in.py:320`) → `eps_factor = jnp.power(EEVPD_wanted / EEVPD, 1/6)` NaN (`:333`).
3. **The NaN safety does not fire.** `nans = Etheta["rejection_rate_nans"] > 0.0` (`:322`) is `False`, because the kernel reported no NaN, so `nan_reject(...)` (`:336-338`) passes the NaN through and `step_size := step_size * eps_factor` (`:357`) becomes NaN.
4. **It is unrecoverable, because the update is multiplicative.** From the next iteration every chain NaNs and the safety *does* fire with `eps_factor = 0.5` — but `NaN * 0.5 = NaN`. The step size never comes back.
5. **The remaining budget is a no-op.** The burn-in kernel reverts every NaN proposal (`laps_burn_in.py:69-70`), so the ensemble stays frozen where it was for the remaining ~975 iterations. The NaN step size flows into phase 2 (`laps.py:254-256, :291`), whose mean acceptance is **exactly 0.000** for all 33 adjusted samples. `laps()` returns finite positions and raises nothing.

Cost in the observed dead runs: **948–972 of 1000** burn-in iterations wasted, plus the entire adjusted phase.

## Two defects

- **D1 (trigger)** — `info.nonans` (`mclmc.py:229-236`) does not cover `kinetic_change` / `energy_change`, so a NaN energy change passes the kernel's own NaN gate and poisons the `psum` ensemble average. The `#969` comment at `laps_burn_in.py:62-70` states this safety is live; it is blind to this route.
- **D2 (why it is fatal rather than a hiccup)** — the step-size update at `laps_burn_in.py:357` is multiplicative, so a non-finite step size is absorbing and the `eps_factor = 0.5` safety cannot recover from it. **D2 is the more serious of the two**: even with D1 fixed, any other route to a NaN step size produces the same silent garbage.

## Evidence

| sweep | n | runs failing the current assertion | runs with non-finite adapted step size |
|---|---|---|---|
| fixed seed-19 dataset, inference key varied | 60 | 8 (**13.3%**) | 8 (13.3%) |
| fresh dataset + key per run | 40 | 6 (15.0%) | 6 (15.0%) |
| numeric perturbations at the test seed | 14 | 6 (42.9%) | 6 (42.9%) |

Across all 115 runs there are **zero intermediate cases** — a run either converges or is dead, and the two signals coincide perfectly:

- converged runs (n=95): phase-2 mean acceptance in **[0.066, 0.972]**
- dead runs (n=20): phase-2 mean acceptance **exactly 0.000**, every time

A second 60-run sweep instrumented for step-size finiteness reproduced this exactly: **60/60 agreement** between "the adapted step size stayed finite" and "the ensemble converged". In the 8 dead runs the NaN first appears at burn-in iteration `[28, 31, 32, 32, 34, 38, 48, 52]`.

This is not a tuning or budget issue: a run with burn-in cut to 100 steps converges fine, so 1000 steps is not marginal. The failure is entirely the NaN death.

## Detection

Non-finiteness of the adapted step size is a perfect predictor in 115/115 runs, and phase-2 mean acceptance is exactly `0.000` in every dead run. Either makes a cheap guard.

## Note on the existing test

`test_laps` asserts `mean(exp(log_scale)) ≈ 1.0`, but the true posterior mean for its dataset is **1.0333** (independent 4-chain NUTS, 1000 warmup + 5000 draws, 0 divergences, R-hat 1.00003, ESS ~1.6e4), so a third of its `atol=1e-1` budget was spent on a wrong target before any sampling. It also reduces the ensemble with `mean(exp(·))`, which no robust statistic can rescue in the failing runs — in those, `median(exp(log_scale)) = 1.76` and `mean(coefs) = 1.48` against truth 1.03 and 3.01.

— 🤖 Blackjax-devs AI TL
